ambari-server/src/main/java/org/apache/ambari/server/controller/internal/DefaultOperatingSystemResourceProvider.java

```diff
@@ -0,0 +1,139 @@
+  public static final String OPERATING_SYSTEM_MPACK_ID = PropertyHelper.getPropertyId("OperatingSystems", "mpack_id");
+
+  private static Set<String> pkPropertyIds = Sets.newHashSet(OPERATING_SYSTEM_OS_TYPE_PROPERTY_ID);
+
+  public static Set<String> propertyIds = Sets.newHashSet(OPERATING_SYSTEM_MPACK_ID,
+      OPERATING_SYSTEM_OS_TYPE_PROPERTY_ID, OPERATING_SYSTEM_REPOS);
+
+  public static Map<Type, String> keyPropertyIds = new HashMap<Type, String>() {
+    {
+      put(Resource.Type.Mpack, OPERATING_SYSTEM_MPACK_ID);
+      put(Resource.Type.DefaultOperatingSystem, OPERATING_SYSTEM_OS_TYPE_PROPERTY_ID);
+    }
+  };
+
+  /**
+   *
+   */
+  @Inject
+  private static Provider<AmbariMetaInfo> m_ambariMetaInfoProvider;
+
+  /**
+   * Constructor.
+   *
+   * @param managementController
+   */
+  protected DefaultOperatingSystemResourceProvider(
+      AmbariManagementController managementController) {
+    super(Resource.Type.DefaultOperatingSystem, propertyIds, keyPropertyIds, managementController);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected Set<String> getPKPropertyIds() {
+    return pkPropertyIds;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public Set<Resource> getResources(Request request, Predicate predicate) throws SystemException,
+      UnsupportedPropertyException, NoSuchResourceException, NoSuchParentResourceException {
+
+    Set<String> requestPropertyIds = getRequestPropertyIds(request, predicate);
+
+    // use a collection which preserves order since JPA sorts the results
+    Set<Resource> results = new LinkedHashSet<>();
+
+    for (Map<String, Object> propertyMap : getPropertyMaps(predicate)) {
+      String mpackIdString = (String) propertyMap.get(OPERATING_SYSTEM_MPACK_ID);
+      Long mpackId = Long.valueOf(mpackIdString);
+
+      Mpack mpack = m_ambariMetaInfoProvider.get().getMpack(mpackId);
+      if (null == mpack) {
+        throw new IllegalArgumentException();
+      }
+
+      RepositoryXml repositoryXml = mpack.getRepositoryXml();
+      for (Os operatingSystem : repositoryXml.getOses()) {
+        Resource resource = toResource(operatingSystem, requestPropertyIds);
+        resource.setProperty(OPERATING_SYSTEM_MPACK_ID, mpack.getResourceId());
+        results.add(resource);
+      }
+    }
+
+    return results;
+  }
+
+  /**
+   * Convert the repository entity to a response resource for serialization.
+   *
+   * @param mpack
+   *          the mpack
+   * @param requestedIds
+   *          the list of requested IDs to use when setting optional properties.
+   * @return the resource to be serialized in the response.
+   */
+  private Resource toResource(Os operatingSystem, Set<String> requestedIds) {
+    Resource resource = new ResourceImpl(Resource.Type.DefaultOperatingSystem);
+    resource.setProperty(OPERATING_SYSTEM_OS_TYPE_PROPERTY_ID, operatingSystem.getFamily());
```

Comment:
We tend to use `org.apache.ambari.server.controller.internal.BaseProvider#setResourceProperty` to set resource values.  This buys us automatic filtering  of the response data based on the user's query.

Reply:
That is true, but in this case, we would probably always want it anyway. Normally, I reserve `setResourceProperty()` for truly optional data or for data that's expensive to calculate.